compiler/rustc_infer/src/infer/error_reporting/suggest.rs

```diff
@@ -412,14 +412,70 @@ impl<'tcx> TypeErrCtxt<'_, 'tcx> {
                 let found_sig =
                     &(self.normalize_fn_sig)(self.tcx.bound_fn_sig(*did2).subst(self.tcx, substs2));
 
-                if self.same_type_modulo_infer(*found_sig, *expected_sig) {
-                    diag.note(
+                diag.note(
                     "different fn items have unique types, even if their signatures are the same",
-                    );
+                );
+
+                if !self.same_type_modulo_infer(*found_sig, *expected_sig)
+                    || !found_sig.is_suggestable(self.tcx, true)
+                    || !expected_sig.is_suggestable(self.tcx, true)
+                    || ty::util::is_intrinsic(self.tcx, *did1)
+                    || ty::util::is_intrinsic(self.tcx, *did2)
+                {
+                    return;
                 }
+
+                let fn_name1 = self.tcx.def_path_str_with_substs(*did1, substs1);
+                let fn_name2 = self.tcx.def_path_str_with_substs(*did2, substs2);
+
+                let (sug1, sug2) = match (expected.is_ref(), found.is_ref()) {
+                    (true, false) => {
+                        let sug1 = format!("&({fn_name1} as {expected_sig})");
+                        let sug2 = format!("{fn_name2} as {found_sig}");
+                        (sug1, sug2)
+                    }
+                    (false, true) => {
+                        let sug1 = format!("{fn_name1} as {expected_sig}");
+                        let sug2 = format!("&({fn_name2} as {found_sig})");
+                        (sug1, sug2)
+                    }
+                    (true, true) => {
+                        let sug1 = format!("&({fn_name1} as {expected_sig})");
+                        let sug2 = format!("&({fn_name2} as {found_sig})");
+                        (sug1, sug2)
+                    }
+                    (false, false) => {
+                        let sug1 = format!("{fn_name1} as {expected_sig}");
+                        let sug2 = format!("{fn_name2} as {found_sig}");
+                        (sug1, sug2)
+                    }
+                };
+                let msg = format!(
+                    "consider casting the expected and found fn items to fn pointers, i.e. `{}` and the following suggestion",
```

Comment:
```suggestion
                    "consider casting both fn items to fn pointers using `as ....`",
```

(fill in the `....` with just the signature they need to be casted to)

And then perhaps using https://doc.rust-lang.org/nightly/nightly-rustc/rustc_errors/diagnostic/struct.Diagnostic.html#method.span_suggestion_hidden

Reply:
Let me explain my thinking here: the phrasing "expected and found fn items" is a bit verbose, and "and the following suggestion" is a bit awkward. Let's avoid both by just explaining the gist of what the user needs to do.

Reply:
Yeah, that explanation is much nicer to read. `--verbose` is my default and I'm realizing how difficult it has been for me to be concise.

Reply:
I've applied the suggestions.